### PR TITLE
Fix expanding home directory for Trainer's `default_root_dir`

### LINKED ISSUE #19059: Expand paths that start with "~" in `Trainer.default_root_dir` and other places
### Description & Motivation

In  #19058 we fixed the issue that `ModelCheckpoint(dirpath="~/checkpoints", ...)` would not expand the home folder in the path. When a user provides a path with this special character at the start, it is unlikely that their intention is to create a folder with that name in the current directory. The special ~ symbol is more commonly used to denote the home folder of the user.

### Pitch

Everywhere in our codebase where we accept a relative path as input from the user, we should expand such paths properly, as early as possible. See #19058 for how to do that.

An immediate example for this is `Trainer(default_root_dir=...)`, but there probably are other places we should double check.

### Alternatives

Keep as is, the user needs to expand the path themselves.

### Additional context

_No response_

cc @carmocca

### PR DESCRIPTION
**Pull request by asingh9530**
## What does this PR do?

If the user provides Trainer(default_root_dir="~/Trainer"), the path doesn't get expanded. Instead, it creates the folder with name ~ in the current working directory. 

Fixes #19059 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19179.org.readthedocs.build/en/19179/

### CODE REVIEW

**Review by Borda — APPROVED** [review]
lets add note to change log

**Review by awaelchli — APPROVED** [review]
Looks great!

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 1]
```diff
@@ -2064,3 +2064,17 @@ def test_init_module_context(monkeypatch):
     with pytest.warns(PossibleUserWarning, match="can't place .* on the device"), trainer.init_module():
         pass
     strategy.tensor_init_context.assert_called_once()
+
+
+def test_expand_home_trainer():
```
[the remark this answers is not in the document]
Comment (asingh9530) [reply tests/tests_pytorch/trainer/test_trainer.py:2069]: @carmocca yes I am working on this

[the remark this answers is not in the document]
Comment (asingh9530) [reply tests/tests_pytorch/trainer/test_trainer.py:2069]: @carmocca github job seems to stuck, could you please restart it ?

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/tests_pytorch/trainer/test_trainer.py:2069]: Done

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 2]
```diff
@@ -2064,3 +2064,13 @@ def test_init_module_context(monkeypatch):
...
+    """Test that the dirpath gets expanded if it contains `~`."""
+    home_root = Path.home()
+
+    checkpoint = Trainer(default_root_dir="~/trainer")
+    assert checkpoint.default_root_dir == str(home_root / "trainer")
+    checkpoint = Trainer(default_root_dir=Path("~/trainer"))
+    assert checkpoint.default_root_dir == str(home_root / "trainer")
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/test_trainer.py:2076]:
```suggestion
    trainer = Trainer(default_root_dir="~/trainer")
    assert trainer.default_root_dir == str(home_root / "trainer")
    trainer = Trainer(default_root_dir=Path("~/trainer"))
    assert trainer.default_root_dir == str(home_root / "trainer")
```

**Review by carmocca — APPROVED** [review]

**Comment by asingh9530** [comment]
@carmocca looks like [test_trainer_compiled_model_test](https://github.com/Lightning-AI/pytorch-lightning/actions/runs/7268423651/job/19804244873?pr=19179) is broken, if possible could you check because it shouldn't fail.
